### PR TITLE
[query] New Shuffler

### PR DESCRIPTION
**Pull request by johnc1231**
PRing for test suite, but it's mostly working. 

Todo:

- [x] Optimization for already sorted tables
- [x] ~~Handle sort by descending~~ (deferred to subsequent PR)
- [x] Handle tables with no partitions

### CODE REVIEW

**Comment by johnc1231** [comment]
I assigned both Tim and Patrick for this since we've been talking about it. It still has prints and stuff in it that I'm using to debug / and experiment, but I think it's now in a place where a review would be welcome.

**Review by patrick-schultz — CHANGES_REQUESTED** [review]
Mostly looks good, though I found it quite hard to read. A few more comments in the big sort method might help.

File: hail/src/main/scala/is/hail/expr/ir/lowering/LowerDistributedSort.scala [hunk 1]
```diff
@@ -78,4 +76,407 @@ object LowerDistributedSort {
...
+
+  def distributedSort(
+    ctx: ExecuteContext,
+    inputStage: TableStage,
+    sortFields: IndexedSeq[SortField],
+    relationalLetsAbove: Map[String, IR],
+    rowTypeRequiredness: RStruct
+  ): TableStage = {
+
+    val oversamplingNum = 1
```
[the remark this answers is not in the document]
Comment (johnc1231) [reply hail/src/main/scala/is/hail/expr/ir/lowering/LowerDistributedSort.scala:111]: Ah, sorry. Yeah I have to increase, right now we are basically not using this.

**Comment by johnc1231** [comment]
@patrick-schultz I addressed your comments, would be happy to take more steps to make code clearer if you see some. Otherwise I'd like if you could check that you're now happy with how `oversamplingNum` is being used.

File: hail/src/main/scala/is/hail/expr/ir/lowering/LowerDistributedSort.scala [hunk 2]
```diff
@@ -176,17 +176,22 @@ object LowerDistributedSort {
...
             val minArray = MakeArray(GetField(aggResults, "min"))
             val maxArray = MakeArray(GetField(aggResults, "max"))
             val tuplesInSortedOrder = tuplesAreSorted(GetField(aggResults, "perPartIntervalTuples"))
-            MakeStruct(Seq(
-              "pivotsWithEndpoints" -> ArrayFunctions.extend(ArrayFunctions.extend(minArray, sortedSamples), maxArray),
-              "isSorted" -> ApplySpecial("land", Seq.empty[Type], Seq(GetField(aggResults, "eachPartSorted"), tuplesInSortedOrder), TBoolean, ErrorIDs.NO_ERROR),
-              "intervalTuple" -> MakeTuple.ordered(Seq(GetField(aggResults, "min"), GetField(aggResults, "max")))
-            ))
+            bindIR(sortedOversampling) { sortedOversampling =>
+              val sortedSampling = ToArray(mapIR(StreamRange(0, ArrayLen(sortedOversampling), I32(oversamplingNum))) { idx =>
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/lowering/LowerDistributedSort.scala:196]: I believe the range you want is `range(oversamplingNum - 1, length, oversamplingNum)`. For example, say `branchingFactor = 4` and `oversamplingNum = 3`. Then we collect `4*3-1 = 11` "oversamples", and we want to extract `4-1` evenly spaced samples:
```
[*, *, 2, *, *, 5, *, *, 8, *, *]
```

**Review by patrick-schultz — CHANGES_REQUESTED** [review]
With one fix to the oversampling logic (and getting tests passing), I'm good with merging this.

**Comment by johnc1231** [comment]
Ok Patrick, back to you. I realized `PruneDeadFields` was disabled in my PR, I renabled it, and had to adjust the prune rule for `AggFold`.

**Review by patrick-schultz — APPROVED** [review]
Sorry I've been so slow reviewing, this is great work!

**Pull request merged**